### PR TITLE
fix(shaker): fix some edge cases related to specific export patterns

### PR DESCRIPTION
**Pull request by jpnelson**
I've tried adopting the shaker method (over extraction) and ran into some edge cases – I've added tests that explain what those edge cases look like.

## Motivation
There are three cases that weren't shaken correctly by the shaker algorithm:
1. Assigning to an export (eg. `var x = module.exports = {...}`)
2. Exporting objects (with computed keys)
3. Exporting objects directly (eg. `module.exports = {...}`) and using this object as a default import elsewhere

## Summary
This fixes it by changing the shaker algorithm to support these cases.

I think so long as we can get these new tests to pass, I'd be happy to take suggestions on exactly how we can implement the fixes. In particular, the biggest change is to support assigning to an export, for example:

```
var x = module.exports = {...}
```

The issue there is that the variable declaration `var x = ...` is marked as not-alive, and so the whole expression is shaken. This fixes it by checking variable declarations, but we could also change the `shake` part of the code to check for alive children. I think that that may have broader implications though.

Very impressed with the shaker algorithm overall, thanks for writing it!

## Test plan
Tested with snapshot tests. I believe for CI to pass, https://github.com/callstack/linaria/pull/950

### CODE REVIEW

**Review by Anber — APPROVED** [review]

**Comment by Anber** [comment]
Looks perfect! Thank you!

File: packages/shaker/src/langs/core.ts [hunk 1]
```diff
@@ -736,17 +736,19 @@ export const visitors: Visitors = {
   /*
    * SequenceExpression
    * It is a special case of expression in which the value of the whole
-   * expression depends only on the last subexpression in the list.
-   * The rest of the subexpressions can be omitted if they don't have dependent nodes.
+   * evaluates to the last subexpression in the list.
+   * In theory, some could be emitted, but we set all the subexpressions
+   * to have dependencies since there could be references in the list.
```
Comment (Anber) [inline packages/shaker/src/langs/core.ts:741]: Could you please provide an example when we need to keep the full list?

Comment (jpnelson) [reply packages/shaker/src/langs/core.ts:741]: I noticed CI was failing so I added that here: https://github.com/callstack/linaria/pull/953

an example is something like:

```
var n = 0;
export default number = (n = 5, n);
```

if we took only the last item in the sequence expression, we'd be left with

```
var n = 0;
export default n;
```

which isn't equivalent.

I created https://github.com/callstack/linaria/pull/953 which adds a proper test case for this, and I noticed CI was broken too. This updates that 🙏

**Comment by jpnelson** [comment]
@Anber would a release be possible after this merges, if CI is green? Thank you again for maintaining this!

File: packages/shaker/src/langs/core.ts (hunk 1 above)
Comment (Anber) [reply packages/shaker/src/langs/core.ts:741]: This example should work properly without changing the shaker, because the last expression is marked as valuable by default and the first is marked as a dependency of the last one, so that both are stay alive.

Comment (Anber) [reply packages/shaker/src/langs/core.ts:741]: Ok, I see now why you changed that behaviour :)
```
	var object = (_object = {}, (0, _defineProperty2.default)(_object, computedKeyName, 'red'), (0, _defineProperty2.default)(_object, "blue", 'blue'), _object);
```

The problem here can be solved by keeping alive all members in sequence, but it isn't the best solution. I'll release the fix today.

Comment (Anber) [reply packages/shaker/src/langs/core.ts:741]: Ok, I see now why you changed that behaviour :)
```
	var object = (_object = {}, (0, _defineProperty2.default)(_object, computedKeyName, 'red'), (0, _defineProperty2.default)(_object, "blue", 'blue'), _object);
```

The problem here can be solved by keeping alive all members in sequence, but it isn't the best solution. I'll release the fix today.
